src/main/kotlin/org/rust/lang/core/types/infer/MemoryCategorization.kt

```diff
@@ -5,140 +5,368 @@
+    fun processExpr(expr: RsExpr): Cmt {
+        val adjustments = expr.inference?.adjustments?.get(expr) ?: emptyList()
+        return processExprAdjustedWith(expr, adjustments.asReversed().iterator())
+    }
 
-        is RsPatBinding -> {
-            if (declaration.mutability.isMut) {
-                Cmt(type, MutabilityCategory.Declared)
-            } else {
-                Cmt(type, MutabilityCategory.Immutable)
+    private fun processExprAdjustedWith(expr: RsExpr, adjustments: Iterator<Adjustment>): Cmt =
+        when (adjustments.nextOrNull()) {
+            is Adjustment.Deref -> {
+                // TODO: overloaded deref
+                processDeref(expr, processExprAdjustedWith(expr, adjustments))
             }
+            else -> processExprUnadjusted(expr)
         }
 
-        is RsSelfParameter -> Cmt(type, MutabilityCategory.valueOf(declaration.mutability))
+    private fun processExprUnadjusted(expr: RsExpr): Cmt =
+        when (expr) {
+            is RsUnaryExpr -> processUnaryExpr(expr)
+            is RsDotExpr -> processDotExpr(expr)
+            is RsIndexExpr -> processIndexExpr(expr)
+            is RsPathExpr -> processPathExpr(expr)
+            is RsParenExpr -> processParenExpr(expr)
+            else -> processRvalue(expr)
+        }
 
-        else -> Cmt(type)
+    private fun processUnaryExpr(unaryExpr: RsUnaryExpr): Cmt {
+        if (!unaryExpr.isDereference) return processRvalue(unaryExpr)
+        val base = unaryExpr.expr ?: return Cmt(unaryExpr, ty = unaryExpr.type)
+        val baseCmt = processExpr(base)
+        return processDeref(unaryExpr, baseCmt)
+    }
+
+    private fun processDotExpr(dotExpr: RsDotExpr): Cmt {
+        if (dotExpr.methodCall != null) {
+            return processRvalue(dotExpr)
+        }
+        val type = dotExpr.type
+        val base = dotExpr.expr
+        val baseCmt = processExpr(base)
+        val fieldName = dotExpr.fieldLookup?.identifier?.text
+        return cmtOfField(dotExpr, baseCmt, fieldName, type)
     }
-}
 
-private fun processParenExpr(parenExpr: RsParenExpr): Cmt =
-    Cmt(parenExpr.type, parenExpr.expr.mutabilityCategory)
+    private fun processIndexExpr(indexExpr: RsIndexExpr): Cmt {
+        val type = indexExpr.type
+        val base = indexExpr.containerExpr ?: return Cmt(indexExpr, ty = type)
+        val baseCmt = processExpr(base)
+        return Cmt(indexExpr, Interior(baseCmt, InteriorIndex), baseCmt.mutabilityCategory.inherit(), type)
+    }
 
-private fun processExpr(expr: RsExpr): Cmt {
-    val adjustments = expr.inference?.adjustments?.get(expr) ?: emptyList()
-    return processExprAdjustedWith(expr, adjustments.asReversed().iterator())
-}
+    private fun processPathExpr(pathExpr: RsPathExpr): Cmt {
+        val type = pathExpr.type
+        val declaration = pathExpr.path.reference.resolve() ?: return Cmt(pathExpr, ty = type)
```

Comment:
Let's use `pathExpr.path.reference.deepResolve()` to make it work with type aliases